Add an example of how to use export helpers in content scripts

wbamberg opened the pull request.

Cobe1 commented:
appreciate the guidance definitely could be more available heading to the
mountains for a brothers anniversary will integrate any needed pre task for
success this week any literature non binding would be greatly appreciated
in up to speed with environment thx

On Aug 19, 2016 3:11 PM, "wbamberg" <notifications@github.com> wrote:

> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/mdn/webextensions-examples/pull/86
> Commit Summary
>
>    - Added example of using export helpers
>    - stop using wrapedJSObject with exportFunction; remove createObjectIn
>
> File Changes
>
>    - *A* export-helpers/README.md
>    <https://github.com/mdn/webextensions-examples/pull/86/files#diff-0>
>    (28)
>    - *A* export-helpers/background.js
>    <https://github.com/mdn/webextensions-examples/pull/86/files#diff-1>
>    (30)
>    - *A* export-helpers/content_scripts/export.js
>    <https://github.com/mdn/webextensions-examples/pull/86/files#diff-2>
>    (24)
>    - *A* export-helpers/icons/LICENSE
>    <https://github.com/mdn/webextensions-examples/pull/86/files#diff-3>
>    (5)
>    - *A* export-helpers/icons/arrow.svg
>    <https://github.com/mdn/webextensions-examples/pull/86/files#diff-4>
>    (1)
>    - *A* export-helpers/manifest.json
>    <https://github.com/mdn/webextensions-examples/pull/86/files#diff-5>
>    (32)
>
> Patch Links:
>
>    - https://github.com/mdn/webextensions-examples/pull/86.patch
>    - https://github.com/mdn/webextensions-examples/pull/86.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mdn/webextensions-examples/pull/86>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/ASqizanyMiVC4ENfA40X1DAWg_HlmN6-ks5qhhv0gaJpZM4Jo270>
> .
>

andymckay commented:
Looks like we need to get something to review this.

andymckay commented:
@rpl could you take a r? please?

rpl commented on export-helpers/README.md:
```diff
@@ -0,0 +1,28 @@
+ of the global `window` object.
+ * defines an object `messenger`, that has a member function `notify()`, and
+ exports that to the page as a property of the global `window` object.
+ * creates an object `messenger2` directly in the page's scope
```
it looks like `messenger2` is not defined (yet?)

rpl commented on export-helpers/README.md:
```diff
@@ -0,0 +1,28 @@
+ 
+The content script then:
+ 
+ * defines a function `notify()` and exports that to the page as a property
```
it could makes sense to explain that the first one  `exportFunction` and the second one `cloneInto with cloneFunctions option`.

wbamberg replied:
If possible, I'd like to get this PR finished off, and we could look at this afterwards?

rpl commented on export-helpers/README.md:
```diff
@@ -0,0 +1,28 @@
+ 
+The content script then:
+ 
+ * defines a function `notify()` and exports that to the page as a property
```
Another use case of `cloneInto` (without the `cloneFunctions` option) that could be reasonable to provide in this example is the following:
- a content script has received from the messaging API (e.g. onMessage or a Port) and wants to forward it to the page into a custom event:
  - on Chrome the object properties are accessible to the webpage javascript code without doing anything
  - on Firefox the object properties are not accessible to the webpage javascript and the objects needs to be explicitly cloned using cloneInto before forwarding as the details of the custom event sent to the page.

The following commit shows a similar change applied to the WebVR Extension to make it to work correctly when running as a WebExtension on Firefox: 

- https://github.com/rpl/WebVR-Extension/commit/1548d4e40a2450fcd13ae8804415970bf9f7bd75

rpl commented on export-helpers/background.js:
```diff
@@ -0,0 +1,30 @@
+Execute content script in the active tab.
+*/
+function loadContentScript() {
+  chrome.tabs.executeScript({
```
If the example is meant to be Firefox specific we could change all the `chrome.*` API calls into `browser.*` API calls
(and given that it demonstrates a Firefox specific feature, it looks like it could be, the only part that could be cross-browser is the suggested "forward an object received through the messaging API to the page in a custom event" scenario).

rpl commented on export-helpers/background.js:
```diff
@@ -0,0 +1,30 @@
+}
+
+/*
+Execute content script in the active tab.
```
Nit, we could indent the inline comment content to make them easier to be spotted.

wbamberg commented:
@rpl , thanks for looking and I'm sorry to be so slow fixing it. Someone was asking about export helpers on IRC today so I thought I should get this finished off.

I fixed your review comments, except https://github.com/mdn/webextensions-examples/pull/86#discussion_r118972444 - if it's OK with you I would like to get this simple example merged now. We could file an issue for this extra case.

I changed things a bit: the content script is injected declaratively into https://github.com/mdn/webextensions-examples/tree/master/export-helpers, which doesn't exist yet, but it will. This should make it a bit easier to run.

rpl commented:
@wbamberg r+ 
Looks great (and I agree, we can add the CustomEvent scenario in a follow up).
